ipvs/ipvs_test.go

```diff
@@ -44,75 +41,73 @@ var (
 	}
 )
 
-func checkDestination(t *testing.T, checkPresent bool, protocol, serviceAddress, realAddress, fwdMethod string) {
-	var (
-		realServerStart bool
-		realServers     []string
-	)
+func lookupFwMethod(fwMethod uint32) string {
 
-	out, err := exec.Command("ipvsadm", "-Ln").CombinedOutput()
-	require.NoError(t, err)
-
-	for _, o := range strings.Split(string(out), "\n") {
-		cmpStr := serviceAddress
-		if protocol == "FWM" {
-			cmpStr = " " + cmpStr
-		}
+	switch fwMethod {
+	case ConnectionFlagMasq:
+		return fwdMethodStrings[0]
+	case ConnectionFlagTunnel:
+		return fwdMethodStrings[1]
+	case ConnectionFlagDirectRoute:
+		return fwdMethodStrings[2]
+	}
+	return ""
+}
 
-		if strings.Contains(o, cmpStr) {
-			realServerStart = true
-			continue
-		}
+func checkDestination(t *testing.T, i *Handle, s *Service, d *Destination, checkPresent bool) {
+	dstFound := false
 
-		if realServerStart {
-			if !strings.Contains(o, "->") {
-				break
-			}
+	dstArray, err := i.GetDestinations(s)
+	require.NoError(t, err)
 
-			realServers = append(realServers, o)
+	for _, dst := range dstArray {
+		if dst.Address.String() == d.Address.String() && dst.Port == d.Port && lookupFwMethod(dst.ConnectionFlags) == lookupFwMethod(d.ConnectionFlags) {
+			dstFound = true
+			break
 		}
 	}
 
-	for _, r := range realServers {
-		if strings.Contains(r, realAddress) {
-			parts := strings.Fields(r)
-			assert.Equal(t, fwdMethod, parts[2])
-			return
+	switch checkPresent {
+	case true: //The test expects the service to be present
+		if !dstFound {
+
+			t.Fatalf("Did not find the service %s in ipvs output", d.Address.String())
+		}
+	case false: //The test expects that the service should not be present
+		if dstFound {
+			t.Fatalf("Did not find the destination %s fwdMethod %s in ipvs output", d.Address.String(), lookupFwMethod(d.ConnectionFlags))
 		}
 	}
 
-	if checkPresent {
-		t.Fatalf("Did not find the destination %s fwdMethod %s in ipvs output", realAddress, fwdMethod)
-	}
 }
 
-func checkService(t *testing.T, checkPresent bool, protocol, schedMethod, serviceAddress string) {
-	out, err := exec.Command("ipvsadm", "-Ln").CombinedOutput()
-	require.NoError(t, err)
+func checkService(t *testing.T, i *Handle, s *Service, checkPresent bool) {
 
-	for _, o := range strings.Split(string(out), "\n") {
-		cmpStr := serviceAddress
-		if protocol == "FWM" {
-			cmpStr = " " + cmpStr
-		}
+	svcArray, err := i.GetServices()
+	require.NoError(t, err)
 
-		if strings.Contains(o, cmpStr) {
-			parts := strings.Split(o, " ")
-			assert.Equal(t, protocol, parts[0])
-			assert.Equal(t, serviceAddress, parts[2])
-			assert.Equal(t, schedMethod, parts[3])
+	svcFound := false
 
-			if !checkPresent {
-				t.Fatalf("Did not expect the service %s in ipvs output", serviceAddress)
-			}
+	for _, svc := range svcArray {
 
-			return
+		if svc.Protocol == s.Protocol && svc.Address.String() == s.Address.String() && svc.Port == s.Port {
+			svcFound = true
```

Comment:
it just that we could have avoided svcFound flag. 
If the service was found {
     if !checkPresent{
                  //error
     }
    //implicit else
     return
}

//outside the loop
if checkPresent{
    //error
}